sled-storage/src/manager.rs

```diff
@@ -0,0 +1,1062 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! The storage manager task
+
+use std::collections::HashSet;
+
+use crate::dataset::{DatasetError, DatasetName};
+use crate::disk::{Disk, DiskError, RawDisk};
+use crate::error::Error;
+use crate::resources::StorageResources;
+use camino::Utf8PathBuf;
+use illumos_utils::zfs::{Mountpoint, Zfs};
+use illumos_utils::zpool::ZpoolName;
+use key_manager::StorageKeyRequester;
+use omicron_common::disk::DiskIdentity;
+use sled_hardware::DiskVariant;
+use slog::{error, info, o, warn, Logger};
+use tokio::sync::{mpsc, oneshot, watch};
+use tokio::time::{interval, Duration, MissedTickBehavior};
+use uuid::Uuid;
+
+// The size of the mpsc bounded channel used to communicate
+// between the `StorageHandle` and `StorageManager`.
+//
+// How did we choose this bound, and why?
+//
+// Picking a bound can be tricky, but in general, you want the channel to act
+// unbounded, such that sends never fail. This makes the channels reliable,
+// such that we never drop messages inside the process, and the caller doesn't
+// have to choose what to do when overloaded. This simplifies things drastically
+// for developers. However, you also don't want to make the channel actually
+// unbounded, because that can lead to run-away memory growth and pathological
+// behaviors, such that requests get slower over time until the system crashes.
+//
+// Our team's chosen solution, and used elsewhere in the codebase, is is to
+// choose a large enough bound such that we should never hit it in practice
+// unless we are truly overloaded. If we hit the bound it means that beyond that
+// requests will start to build up and we will eventually topple over. So when
+// we hit this bound, we just go ahead and panic.
+//
+// Picking a channel bound is hard to do empirically, but practically, if
+// requests are mostly mutating task local state, a bound of 1024 or even 8192
+// should be plenty. Tasks that must perform longer running ops can spawn helper
+// tasks as necessary or include their own handles for replies rather than
+// synchronously waiting. Memory for the queue can be kept small with boxing of
+// large messages.
+//
+// Here we start relatively small so that we can evaluate our choice over time.
+const QUEUE_SIZE: usize = 256;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum StorageManagerState {
+    WaitingForKeyManager,
+    QueueingDisks,
+    Normal,
+}
+
+enum AddDiskResult {
+    DiskInserted,
+    DiskAlreadyInserted,
+    DiskQueued,
+}
+
+impl AddDiskResult {
+    fn disk_inserted(&self) -> bool {
+        match self {
+            AddDiskResult::DiskInserted => true,
+            _ => false,
+        }
+    }
+}
+
+impl From<bool> for AddDiskResult {
```

Comment:
`AddDiskResult` seems a lot clearer than `bool` - do we need this `From` impl at all? Or could places where we're currently relying on it be changed to use `AddDiskResult`?

Reply:
Good idea. Done in [cfc3ef7](https://github.com/oxidecomputer/omicron/pull/4332/commits/cfc3ef7c732afdfbf476ac4a62b64471c697bfbb)